src/annotator/anchoring/range-util.js

```diff
@@ -0,0 +1,227 @@
+ */
+function getNodePosition(node) {
+  let pos = 0;
+  let tmp = node;
+  while (tmp) {
+    if (tmp.nodeName === node.nodeName) {
+      pos += 1;
+    }
+    tmp = tmp.previousSibling;
+  }
+  return pos;
+}
+
+/**
+ * A simple XPath evaluator using only standard DOM methods which can
+ * evaluate queries of the form /tag[index]/tag[index].
+ */
+function simpleXPathPure(nodes, relativeRoot) {
+  function getPathSegment(node) {
+    const name = getNodeName(node);
+    const pos = getNodePosition(node);
+    return `${name}[${pos}]`;
+  }
+
+  let rootNode = relativeRoot;
+
+  function getPathTo(node) {
+    let xpath = '';
+    let elem = node;
+    while (elem !== rootNode) {
+      if (!elem) {
+        throw new Error(
+          'Called getPathTo on a node which was not a descendant of @rootNode. ' +
+            rootNode
+        );
+      }
+      xpath = getPathSegment(elem) + '/' + xpath;
+      elem = elem.parentNode;
+    }
+    xpath = '/' + xpath;
+    xpath = xpath.replace(/\/$/, '');
+    return xpath;
+  }
+  const paths = nodes.map((index, node) => {
+    return getPathTo(node);
+  });
+  return paths.get();
+}
+
+/**
+ * Public: Flatten a nested array structure
+ */
+function flatten(array) {
+  const flatten = ary => {
+    let flat = [];
+    ary.forEach(el => {
+      if (el && $.isArray(el)) {
+        flat = flat.concat(flatten(el));
+      } else {
+        flat = flat.concat(el);
+      }
+    });
+    return flat;
+  };
+  return flatten(array);
+}
+
+/**
+ * Public: Finds all text nodes within the elements in the current collection.
+ * Returns a new jQuery collection of text nodes.
+ */
+export function getTextNodes(jq) {
+  const getTextNodes = node => {
+    if (node && node.nodeType !== Node.TEXT_NODE) {
+      let nodes = [];
+      if (node.nodeType !== Node.COMMENT_NODE) {
+        [...node.childNodes].forEach(child => {
+          nodes.push(getTextNodes(child));
+        });
+      }
+      return nodes;
+    } else {
+      return node;
+    }
+  };
+  return jq.map((index, node) => {
+    return flatten(getTextNodes(node));
+  });
+}
+
+/**
+ * Public: determine the last text node inside or before the given node
+ */
+export function getLastTextNodeUpTo(n) {
+  switch (n.nodeType) {
+    case Node.TEXT_NODE:
+      return n; // We have found our text node.
+    case Node.ELEMENT_NODE:
+      // This is an element, we need to dig in
+      if (n.lastChild) {
+        // Does it have children at all?
+        const result = getLastTextNodeUpTo(n.lastChild);
+        if (result) {
+          return result;
+        }
+      }
+  }
+  // Could not find a text node in current node, go backwards
+  const prev = n.previousSibling;
+  if (prev) {
+    // eslint-disable-next-line no-unused-vars
+    return getLastTextNodeUpTo(prev);
+  } else {
+    return null;
+  }
+}
+
+/**
+ * Public: determine the first text node in or after the given jQuery node.
```

Comment:
The `Public: ` part can be dropped from the comments - this is implied by the `export` keyword.